infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/msgHandlers/ReplicationHandlerMethods.java

```diff
@@ -0,0 +1,192 @@
+                MethodHandle mh = caller.findVirtual(server.getClass(), method.getName(), mt);
+
+                // Note: mtt captures the return type as well as the argument types of the annotated handler
+                // method. These should be consistent with the handle method described by the HandlerMethod
+                // functional interface.
+                MethodType mtt = mh.type().dropParameterTypes(0, 1);
+                h = (HandlerMethod) LambdaMetafactory.metafactory(caller, "handle",
+                        MethodType.methodType(HandlerMethod.class, server.getClass()),
+                        mtt, mh, mtt).getTarget().bindTo(server).invoke();
+            }
+
+            String payloadSting = annotation.requestType().equals(RequestPayloadMsg.PayloadCase.NONE) ?
+                    annotation.responseType().toString() : annotation.requestType().toString();
+
+            // Install pre-conditions on handler
+            final HandlerMethod handler = generateConditionalHandler(payloadSting, h);
+            // Install the handler in the map
+            if (!annotation.requestType().equals(RequestPayloadMsg.PayloadCase.NONE)) {
+                requestHandlerMap.put(annotation.requestType(), handler);
+            } else {
+                responseHandlerMap.put(annotation.responseType(), handler);
+            }
+        } catch (Throwable e) {
+            throw new UnrecoverableCorfuError("Exception during request handler registration", e);
+        }
+    }
+
+    /**
+     * Generate a conditional handler, which instruments the handler with metrics if
+     * configured as enabled and checks whether the server is shutdown and ready.
+     * @param payloadType          The type the message is being handled for.
+     * @param handler       The {@link HandlerMethod} which handles the message.
+     * @return              A new {@link HandlerMethod} which conditionally executes the
+     *                      handler based on preconditions (whether the server is shutdown/ready).
+     */
+    private HandlerMethod generateConditionalHandler(@NonNull final String payloadType,
+                                                                           @NonNull final HandlerMethod handler) {
+        // Generate a timer name based on the LR message type
+        String timerName = getTimerName(payloadType);
+        // Register the handler. Depending on metrics collection configuration by MetricsUtil,
+        // handler will be instrumented by the metrics context.
+        return (req, res, r) -> MicroMeterUtils.time(() -> handler.handle(req, res, r), timerName);
+    }
+
+    // Create a timer using cached timer name for the corresponding type
+    private String getTimerName(@Nonnull String type) {
+        timerNameCache.computeIfAbsent(type,
+                aType -> ("corfu.infrastructure.message-handler." +
+                        aType.toLowerCase()));
+        return timerNameCache.get(type);
```

Comment:
I think this will not automatically work until we add LR handlers to `MicroMeterUtils.java`

Reply:
I did not find any specific handlers...all seemed generic.
Please let me know if I missed something.